rudder-web/src/main/webapp/javascript/rudder/angular/dataSource.js

```diff
@@ -0,0 +1,186 @@
+var app = angular.module("datasource", []);
+
+app.directive('focusOn', function() {
+  return function(scope, elem, attr) {
+    scope.$on('focusOn', function(e, name) {
+      if(name === attr.focusOn) {
+        elem[0].focus();
+      }
+    });
+  };
+});
+app.factory('focus', function ($rootScope, $timeout) {
+  return function(name) {
+    $timeout(function (){
+      $rootScope.$broadcast('focusOn', name);
+    });
+  }
+});
+
+app.controller("datasourceCtrl", ['$scope', '$timeout', 'orderByFilter','$http', function($scope,$timeout,orderBy,$http,focus) {
+  $scope.forms = {};
```

Comment:
It is used directly in the html code :
<form [...] name="forms.datasourceForm">
It gives access to the form state outside it.